src/core/remio.c

```diff
@@ -0,0 +1,709 @@
+    REMIO_STATE_COMPLETE,   /**< The I/O request was completed by the backend VM but not yet
+                                   completed on frontend VM's eyes */
+};
+
+/**
+ * @union remio_cpu_msg_data
+ * @brief This union represents the Remote I/O CPU message data
+ */
+union remio_cpu_msg_data {
+    struct {
+        uint8_t id;        /**< Remote I/O ID */
+        uint8_t cpu_id;    /**< Frontend CPU ID */
+        uint8_t vcpu_id;   /**< Frontend vCPU ID */
+        uint8_t interrupt; /**< Interrupt ID */
+    };
+    uint64_t raw;          /**< Raw data */
+};
+
+/**
+ * @struct remio_request
+ * @brief This structure contains the information of a Remote I/O request
+ */
+struct remio_request {
+    vaddr_t addr;               /**< Address of the accessed MMIO Register */
+    unsigned long access_width; /**< Access width */
+    unsigned long op;           /**< MMIO operation type (read or write) */
+    unsigned long value;        /**< Value to be written or read */
+    unsigned long reg;          /**< vCPU resgiter used during the MMIO access */
+    enum REMIO_STATE state;     /**< I/O request state */
+};
+
+/**
+ * @struct remio_request_event
+ * @brief This structure contains the information of a Remote I/O request event
+ * @note  This is used to only advertise the backend VM that a new I/O request is available
+ *        and the backend VM can ask for it, avoiding the need to iterate over all the
+ *        CPUs and vCPUs slots to find the next I/O request to be processed
+ */
+struct remio_request_event {
+    node_t node;      /** Node */
+    cpuid_t cpu_id;   /** CPU ID of the frontend VM that issued the I/O request */
+    vcpuid_t vcpu_id; /** vCPU ID of the frontend VM that issued the I/O request */
+};
+
+/**
+ * @struct remio_device_config
+ * @brief This structure holds the static information regarding a Remote I/O device
+ */
+struct remio_device_config {
+    struct {
+        cpuid_t cpu_id;    /**< Backend VM CPU ID */
+        vmid_t vm_id;      /**< Backend VM ID */
+        irqid_t interrupt; /**< Backend interrupt ID */
+    } backend;
+    struct {
+        cpuid_t cpu_id;    /**< Frontend VM CPU ID */
+        vmid_t vm_id;      /**< Frontend VM ID */
+        irqid_t interrupt; /**< Frontend interrupt ID */
+    } frontend;
+};
+
+/**
+ * @struct remio_device
+ * @brief This structure comprises all the information needed about a Remote I/O device
+ */
+struct remio_device {
+    node_t node;                       /**< Node */
+    remio_id_t id;                     /**< Remote I/O device ID */
+    struct remio_device_config config; /**< Remote I/O device configuration */
+    struct list request_event_list;    /**< List of pending I/O requests events */
+};
+
+/** List of Remote I/O devices */
+struct list remio_device_list;
+
+/** Array of Remote I/O requests */
+struct remio_request remio_requests[REMIO_CPU_NUM][REMIO_VCPU_NUM];
```

Comment:
More, I don't really like the way were are allocating requests. I would suggest changing this structure. Since each request is associated with an event, we could merge the two concepts and only free the object from the pool once the request was completed. 

Then instead of indexing requests with the cpu_id/vcpu_id combination we would have a dynamically allocated id for events/requests that would be passed to the backend for later identification on hypercalls. Also associated with it, we would have a "backend owner VM id" that the hypercall handler would use to check that that event/request is indeed being hanlded by that VM.

Reply:
> This seems to be wasting memory since for a given CPU there is at most one vcpu.

Yes, you're right. However, I implemented it this way to allow for potential future enhancements in Bao, such as exploring more advanced configurations. These could include 1-to-many mappings between physical and virtual CPUs, or sharing the same vCPU across multiple VMs. In the former case, the vCPU ID would correspond to the actual vCPU, while in the latter, it would be associated with the VM ID.

Reply:
> More, I don't really like the way were are allocating requests. I would suggest changing this structure. Since each request is associated with an event, we could merge the two concepts and only free the object from the pool once the request was completed.

From my tests, I attempted to merge the two concepts, but based on my experience, this introduces higher costs in terms of both code complexity and performance. First, since the request event is associated with a specific remote I/O device, when the backend requests a new event, it doesn’t need to iterate through all vCPUs to retrieve it. Instead, it can simply pop the oldest element from the list.

Additionally, the issue of vCPU prioritization that I mentioned earlier still persists when merging the two concepts. Furthermore, as the list is already protected by a spinlock, there’s no need for additional synchronization primitives, which results in simpler code.

Reply:
> Then instead of indexing requests with the cpu_id/vcpu_id combination we would have a dynamically allocated id for events/requests that would be passed to the backend for later identification on hypercalls.

But how do you locate the request? Are you suggesting iterating over all vCPUs associated with a frontend VM? The same issue I mentioned earlier would persist in this approach.

I do agree that we could dynamically create a unique identifier to replace the CPU and vCPU IDs. Perhaps we could apply a basic mask and concatenate the two numbers into a single unsigned long variable, generating a unique identifier that could then be used to index the request.

> Also associated with it, we would have a "backend owner VM id" that the hypercall handler would use to check that that event/request is indeed being hanlded by that VM.

I think I may have misunderstood what you referred to as the "backend owner VM ID." However, the current implementation already verifies that the backend VM performing the hypercall is the owner of the remote I/O backend device. Since the event request list is associated with each remote I/O device, we ensure that only the corresponding backend interacts with the event request.

Reply:
The logic of handling the request queue would be exactly the same. The only thing that would fundamentally change in my suggestion would be the way you allocate events/requests and the way the backend identifies the specific event/request on an hypercall.

Reply:
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/3e9940ee1b0566c3ad0b9049454f3e61f7adbc1b

Reply:
I still think we dont need the event array in the device. We can directly access it via the objpool. We just need some check to make sure that event belongs to the VM calling the ask hypercall and that that VM is the backend for the device associated with that event.

Reply:
Solved [here](https://github.com/bao-project/bao-hypervisor/pull/176/commits/77ff8e893f35c268097812e29bcc0463c8796d43)